res/res_pjsip_refer.c

```diff
@@ -786,6 +791,498 @@ static void refer_blind_callback(struct ast_channel *chan, struct transfer_chann
+		ast_refer_var_unref_current(iter)) {
+		if (!is_refer_var_blocked(name)) {
+			ast_sip_add_header(tdata, name, value);
+		}
+	}
+	ast_refer_var_iterator_destroy(iter);
+
+	return PJSIP_SC_OK;
+}
+
+struct refer_out_of_dialog {
+	pjsip_dialog *dlg;
+	int authentication_challenge_count;
+};
+
+/*! \brief REFER Out-of-dialog module, used to attach session data structure to subscription */
+static pjsip_module refer_out_of_dialog_module = {
+	.name = { "REFER Out-of-dialog Module", 26 },
+	.id = -1,
+	.on_tx_request = refer_on_tx_request,
+	/* Ensure that we are called after res_pjsp_nat module and before transport priority */
+	.priority = PJSIP_MOD_PRIORITY_TSX_LAYER - 4,
+};
+
+/*! \brief Helper function which returns the name-addr of the Refer-To header or NULL */
+static pjsip_uri *get_refer_to_uri(pjsip_tx_data *tdata)
+{
+	const pj_str_t REFER_TO = { "Refer-To", 8 };
+	pjsip_generic_string_hdr *refer_to;
+	pjsip_uri *parsed_uri;
+
+	if (!(refer_to = pjsip_msg_find_hdr_by_name(tdata->msg, &REFER_TO, NULL))
+		|| !(parsed_uri = pjsip_parse_uri(tdata->pool, refer_to->hvalue.ptr, refer_to->hvalue.slen, 0))
+		|| (!PJSIP_URI_SCHEME_IS_SIP(parsed_uri) && !PJSIP_URI_SCHEME_IS_SIPS(parsed_uri))) {
+		return NULL;
+	}
+
+	return parsed_uri;
+}
+
+static pj_status_t refer_on_tx_request(pjsip_tx_data *tdata) {
+	RAII_VAR(struct ast_str *, refer_to_str, ast_str_create(PJSIP_MAX_URL_SIZE), ast_free_ptr);
+	const pj_str_t REFER_TO = { "Refer-To", 8 };
+	pjsip_generic_string_hdr *refer_to_hdr;
+	pjsip_dialog *dlg;
+	struct refer_data *refer_data;
+	pjsip_uri *parsed_uri;
+	pjsip_sip_uri *refer_to_uri;
+
+	/*
+	 * If this is a request in response to a 401/407 Unauthorized challenge, the
+	 * Refer-To URI has been rewritten already, so don't attempt to re-write it again.
+	 * Checking for presence of the Authorization header is not an ideal solution. We do this because
+	 * there exists some race condition where this dialog is not the same as the one used
+	 * to send the original request in which case we don't have the correct refer_data.
+	 */
+	if (!refer_to_str
+		|| pjsip_msg_find_hdr(tdata->msg, PJSIP_H_AUTHORIZATION, NULL)
+		|| !(dlg = pjsip_tdata_get_dlg(tdata))
+		|| !(refer_data = pjsip_dlg_get_mod_data(dlg, refer_out_of_dialog_module.id))
+		|| !refer_data->to_self
+		|| !(parsed_uri = get_refer_to_uri(tdata))) {
+		goto out;
+	}
+	refer_to_uri = pjsip_uri_get_uri(parsed_uri);
+	ast_sip_rewrite_uri_to_local(refer_to_uri, tdata);
+
+	pjsip_uri_print(PJSIP_URI_IN_CONTACT_HDR, parsed_uri, ast_str_buffer(refer_to_str), ast_str_size(refer_to_str));
+	refer_to_hdr = pjsip_msg_find_hdr_by_name(tdata->msg, &REFER_TO, NULL);
+	pj_strdup2(tdata->pool, &refer_to_hdr->hvalue, ast_str_buffer(refer_to_str));
+
+out:
+	return PJ_SUCCESS;
+}
+
+static int refer_unreference_dialog(void *obj)
+{
+	struct refer_out_of_dialog *data = obj;
+
+	/* This is why we keep the dialog on the subscription. When the subscription
```

Comment:
How do the underlying PJSIP implementations that use dialogs (such as the outgoing SUBSCRIBE) handle this? Do they use a group lock with destruction callback?

Reply:
I'm not sure if I understand correctly, but yes, e.g. `evsub.c` in PJSIP uses a group lock and in the destruction handler of the group lock, it calls `pjsip_dlg_dec_session`.

Reply:
Is there other pattern of usage with the way you've implemented this already in the tree?